test/jasmine/tests/sankey_test.js

```diff
@@ -908,208 +908,252 @@ function assertNoLabel() {
-        }, {
-            'node': 1,
-            'name': 'node1'
-        }, {
-            'node': 2,
-            'name': 'node2'
-        }, {
-            'node': 3,
-            'name': 'node3'
-        }, {
-            'node': 4,
-            'name': 'node4'
-        }],
-        'links': [{
-            'source': 0,
-            'target': 2,
-            'value': 2
-        }, {
-            'source': 1,
-            'target': 2,
-            'value': 2
-        }, {
-            'source': 1,
-            'target': 3,
-            'value': 2
-        }, {
-            'source': 0,
-            'target': 4,
-            'value': 2
-        }, {
-            'source': 2,
-            'target': 3,
-            'value': 2
-        }, {
-            'source': 2,
-            'target': 4,
-            'value': 2
-        }, {
-            'source': 3,
-            'target': 4,
-            'value': 4
-        }]
-    };
-    var sankey;
-    var graph;
-    var margin = {
-        top: 10,
-        right: 10,
-        bottom: 10,
-        left: 10
-    };
-    var width = 1200 - margin.left - margin.right;
-    var height = 740 - margin.top - margin.bottom;
-
-    beforeEach(function() {
-        sankey = d3sankey
-        .sankey()
-        .nodeWidth(36)
-        .nodePadding(10)
-        .nodes(data.nodes)
-        .links(data.links)
-        .size([width, height])
-        .iterations(32);
-
-        graph = sankey();
-    });
-
-    function checkArray(key, result) {
-        var value = graph.nodes.map(function(obj) {
+describe('sankey layout generators', function() {
+    function checkArray(arr, key, result) {
+        var value = arr.map(function(obj) {
             return obj[key];
         });
         expect(value).toEqual(result, 'invalid property named ' + key);
     }
 
-    function checkRoundedArray(key, result) {
-        var value = graph.nodes.map(function(obj) {
+    function checkRoundedArray(arr, key, result) {
+        var value = arr.map(function(obj) {
             return Math.round(obj[key]);
         });
         expect(value).toEqual(result, 'invalid property named ' + key);
     }
 
-    it('controls the width of nodes', function() {
-        expect(sankey.nodeWidth()).toEqual(36, 'incorrect nodeWidth');
-    });
+    function moveNode(sankey, graph, nodeIndex, delta) {
+        var node = graph.nodes[nodeIndex];
+        var pos0 = [node.x0, node.y0];
+        var pos1 = [node.x1, node.y1];
 
-    it('controls the padding between nodes', function() {
-        expect(sankey.nodePadding()).toEqual(10, 'incorrect nodePadding');
-    });
+        // Update node's position
+        node.x0 += delta[0];
+        node.x1 += delta[0];
+        node.y0 += delta[1];
+        node.y1 += delta[1];
 
-    it('controls the padding between nodes', function() {
-        expect(sankey.nodePadding()).toEqual(10, 'incorrect nodePadding');
-    });
+        // Update links
+        var updatedGraph = sankey.update(graph);
```

Comment:
Actually, can we make these test use `restyle` instead of (the internal) `sankey.update`?

Reply:
> Actually, can we make these tests use `restyle` instead of (the internal) `sankey.update`?

I don't think we can do that right now because we don't have any attributes to control the position of the nodes (so no `restyle` call can change the position of the nodes). Right now, `sankey.update` is only called when nodes are dragged around with the mouse.

Maybe tests using `restyle` could be included in a future PR implementing attributes `node.x` and `node.y`? Or maybe I should add those attributes here right now?

Reply:
Ha I see. Good point.

Can you add (just) one `restyle` test that update circular to non-circular node and links data? Or better yet, one `react` test that goes from one circular mock to a non-circular one?

Reply:
One last TODO :arrow_up: